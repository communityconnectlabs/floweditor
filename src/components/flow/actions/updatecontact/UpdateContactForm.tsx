import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors } from 'components/flow/actions/helpers';
import {
  initializeForm,
  sortFieldsAndProperties,
  stateToAction,
  UpdateContactFormState
} from 'components/flow/actions/updatecontact/helpers';
import { ActionFormProps } from 'components/flow/props';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TextInputElement from 'components/form/textinput/TextInputElement';
import { getContactProperties } from 'components/helpers';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import { Types } from 'config/interfaces';
import { ContactProperties } from 'flowTypes';
import * as React from 'react';
import { Asset, AssetType, updateAssets } from 'store/flowContext';
import * as mutators from 'store/mutators';
import { mergeForm, ValidationFailure } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { shouldRequireIf, validate } from 'store/validators';

import styles from './UpdateContactForm.module.scss';
import i18n from 'config/i18n';

export default class UpdateContactForm extends React.Component<
  ActionFormProps,
  UpdateContactFormState
> {
  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: ActionFormProps) {
    super(props);

    this.state = initializeForm(this.props.nodeSettings, this.props.assetStore);

    bindCallbacks(this, {
      include: [/^get/, /^on/, /^handle/]
    });
  }

  private handleUpdate(
    keys: {
      type?: Types;
      name?: string;
      channel?: Asset;
      language?: Asset;
      field?: Asset;
      fieldValue?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<UpdateContactFormState> = {};

    if (keys.hasOwnProperty('type')) {
      updates.type = keys.type;
    }

    if (keys.hasOwnProperty('name')) {
      updates.name = { value: keys.name };
    }

    if (keys.hasOwnProperty('channel')) {
      updates.channel = validate('Channel', keys.channel, [shouldRequireIf(submitting)]);
    }

    if (keys.hasOwnProperty('language')) {
      updates.language = validate('Language', keys.language, [shouldRequireIf(submitting)]);
    }

    if (keys.hasOwnProperty('field')) {
      updates.field = { value: keys.field };
    }

    if (keys.hasOwnProperty('fieldValue')) {
      updates.fieldValue = { value: keys.fieldValue, validationFailures: [] };
    }

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private handlePropertyChange(selected: Asset[]): boolean {
    const selection = selected[0];
    if (selection.type === AssetType.ContactProperty) {
      switch (selection.id) {
        case ContactProperties.Name:
          return this.handleUpdate({
            field: selection,
            type: Types.set_contact_name,
            name: ''
          });
        case ContactProperties.Language:
          return this.handleUpdate({
            field: selection,
            type: Types.set_contact_language
          });
        case ContactProperties.Channel:
          return this.handleUpdate({
            field: selection,
            type: Types.set_contact_channel
          });
      }
    }
    return this.handleUpdate({
      type: Types.set_contact_field,
      field: selection,
      fieldValue: ''
    });
  }

  private handleChannelUpdate(selection: Asset[], submitting = false): boolean {
    return this.handleUpdate({ channel: selection[0] }, submitting);
  }

  private handleLanguageUpdate(selection: Asset[], submitting = false): boolean {
    return this.handleUpdate({ language: selection[0] }, submitting);
  }

  private handleFieldValueUpdate(fieldValue: string): boolean {
    return this.handleUpdate({ fieldValue, name: '' });
  }

  private handleNameUpdate(name: string): boolean {
    return this.handleUpdate({ name, fieldValue: '' });
  }

  private onUpdated(dispatch: DispatchWithState, getState: GetState): void {
    const {
      flowContext: { assetStore }
    } = getState();

    if (this.state.field.value.type === AssetType.Field) {
      dispatch(updateAssets(mutators.addAssets('fields', assetStore, [this.state.field.value])));
    }
  }

  public handleFieldAdded(field: Asset): void {
    // update our store with our new group
    this.props.addAsset('fields', field);
    this.handlePropertyChange([field]);
  }

  private handleSave(): void {
    let valid = this.state.valid;

    // check if language required
    if (this.state.type === Types.set_contact_language) {
      valid = this.handleLanguageUpdate([this.state.language.value], true) && valid;
    }

    // check if channel required
    if (this.state.type === Types.set_contact_channel) {
      valid = this.handleChannelUpdate([this.state.channel.value], true) && valid;
    }

    /*        if (this.state.type === Types.set_contact_field) {
            const fieldValue = mergePersistantFailures(this.state.fieldValue);
            if (fieldValue.validationFailures.length > 0) {
                valid = this.state.
                this.setState({ fieldValue });
            }
        }

        if (this.state.type === Types.set_contact_name) {
            const name = mergePersistantFailures(this.state.name);
            if (name.validationFailures.length > 0) {
                valid = false;
                this.setState({ name });
            }
        }
*/
    if (valid) {
      // do the saving!
      this.props.updateAction(stateToAction(this.props.nodeSettings, this.state), this.onUpdated);
      this.props.onClose(true);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  /**
   * The value widget varies for the action type
   */
  private getValueWidget(): JSX.Element {
    if (this.state.type === Types.set_contact_channel) {
      return (
        <AssetSelector
          key="select_channel"
          name="Channel"
          placeholder="Select the channel to use for this contact"
          assets={this.props.assetStore.channels}
          entry={this.state.channel}
          searchable={true}
          valueClearable={true}
          onChange={this.handleChannelUpdate}
        />
      );
    }

    if (this.state.type === Types.set_contact_language) {
      return (
        <AssetSelector
          key="select_language"
          name="Language"
          placeholder="Select the language to use for this contact"
          assets={this.props.assetStore.languages}
          entry={this.state.language}
          searchable={true}
          valueClearable={true}
          onChange={this.handleLanguageUpdate}
          shouldExclude={(asset: Asset) => asset.id === 'base'}
        />
      );
    } else if (this.state.type === Types.set_contact_name) {
      return (
        <TextInputElement
          name="Name"
          placeholder="Enter a new name for the contact"
          onFieldFailures={(persistantFailures: ValidationFailure[]) => {
            const name = { ...this.state.name, persistantFailures };
            this.setState({
              name,
              valid: this.state.valid && !hasErrors(name)
            });
          }}
          onChange={this.handleNameUpdate}
          entry={this.state.name}
          autocomplete={true}
          focus={true}
        />
      );
    } else {
      return (
        <TextInputElement
          name="Field Value"
          placeholder={`Enter a new value for ${this.state.field.value.name}`}
          onFieldFailures={(persistantFailures: ValidationFailure[]) => {
            const fieldValue = { ...this.state.fieldValue, persistantFailures };
            this.setState({
              fieldValue,
              valid: this.state.valid && !hasErrors(fieldValue)
            });
          }}
          onChange={this.handleFieldValueUpdate}
          entry={this.state.fieldValue}
          autocomplete={true}
          focus={true}
        />
      );
    }
  }

  public handleCreateAssetFromInput(input: string): any {
    return { label: input, value_type: 'text' };
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />

        <p>Select what to update</p>
        <AssetSelector
          name="Contact Field"
          assets={this.props.assetStore.fields}
          additionalOptions={getContactProperties(this.context.config.flowType)}
          entry={this.state.field}
          searchable={true}
          sortFunction={sortFieldsAndProperties}
          onChange={this.handlePropertyChange}
          // Fields can be created on the fly
          createPrefix="Create Contact Field: "
          createAssetFromInput={this.handleCreateAssetFromInput}
          onAssetCreated={this.handleFieldAdded}
        />

        <div className={styles.value}>{this.getValueWidget()}</div>
      </Dialog>
    );
  }
}
