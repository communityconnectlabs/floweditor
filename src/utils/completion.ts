import { AssetStore, CompletionOption } from 'store/flowContext';

export interface CompletionAssets {
  assetStore: AssetStore;
  schema: CompletionSchema;
}

export interface CompletionProperty {
  key: string;
  help: string;
  type: string;
}

export interface CompletionType {
  name: string;

  key_source?: string;
  property_template?: CompletionProperty;
  properties?: CompletionProperty[];
}

export interface CompletionSchema {
  types: CompletionType[];
  root: CompletionProperty[];
}

export const getFunctions = (functions: CompletionOption[], query: string): CompletionOption[] => {
  return functions.filter((option: CompletionOption) => {
    if (option.signature) {
      return option.signature.indexOf(query) === 0;
    }
    return false;
  });
};

/**
 * Takes a dot query and returns the completions options at the current level
 * @param dotQuery query such as "results.field_name.ca"
 */
export const getCompletions = (
  completion: CompletionAssets,
  dotQuery: string
): CompletionOption[] => {
  const parts = (dotQuery || '').split('.');
  let currentProps: CompletionProperty[] = completion.schema.root;

  let prefix = '';
  let part = '';
  while (parts.length > 0) {
    part = parts.shift();
    if (part) {
      // eslint-disable-next-line
      const nextProp = currentProps.find((prop: CompletionProperty) => prop.key === part);
      if (nextProp) {
        // eslint-disable-next-line
        const nextType = completion.schema.types.find(
          (type: CompletionType) => type.name === nextProp.type
        );
        if (nextType && nextType.properties) {
          currentProps = nextType.properties;
          prefix += part + '.';
        } else if (nextType && nextType.property_template) {
          prefix += part + '.';
          const template = nextType.property_template;
          currentProps = Object.keys(completion.assetStore[nextType.name].items).map(
            (key: string) => ({
              key: template.key.replace('{key}', key),
              help: template.help.replace('{key}', key),
              type: template.type
            })
          );
        } else {
          // eslint-disable-next-line
          currentProps = currentProps.filter((prop: CompletionProperty) =>
            prop.key.startsWith(part.toLowerCase())
          );
          break;
        }
      } else {
        // eslint-disable-next-line
        currentProps = currentProps.filter((prop: CompletionProperty) =>
          prop.key.startsWith(part.toLowerCase())
        );
        break;
      }
    }
  }

  return currentProps.map((prop: CompletionProperty) => {
    const name =
      prop.key === '__default__' ? prefix.substr(0, prefix.length - 1) : prefix + prop.key;
    return { name, summary: prop.help };
  });
};

export const getCompletionName = (option: CompletionOption): string => {
  return option.name || option.signature.substr(0, option.signature.indexOf('('));
};

export const getCompletionSignature = (option: CompletionOption): string => {
  return option.signature.substr(option.signature.indexOf('('));
};
