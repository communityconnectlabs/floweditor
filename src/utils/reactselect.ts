import { StylesConfig } from 'react-select';

export const large = {
  placeholder: (styles: StylesConfig, state: any) => {
    return { ...styles, color: '#ddd', marginLeft: '3px' };
  },
  singleValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '3px' };
  },
  multiValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '0px' };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  }
};

export const messageStyle = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      borderBottom: 'none',
      borderBottomRightRadius: '0px',
      borderBottomLeftRadius: '0px',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  }
};

export const getErroredSelect = (baseControl: any): any => {
  return {
    control: (styles: StylesConfig) => ({
      ...styles,
      ...baseControl,
      borderColor: 'tomato',
      boxShadow: '0 0 0 3px rgba(255,196,186,0.5) !important;',
      '&:hover': {
        borderColor: 'tomato'
      }
    })
  };
};

export const small = {
  control: (styles: StylesConfig) => ({
    ...styles,
    borderColor: '#ccc',
    boxShadow: 'none',
    '&:hover': {
      borderColor: '#ccc'
    },
    maxHeight: '28px',
    minHeight: '28px',
    fontSize: '12px',
    lineHeight: '12px'
  }),
  menuList: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 200
  }),
  option: (styles: StylesConfig) => ({
    ...styles,
    fontSize: 12
  }),
  singleValue: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 28,
    lineHeight: '13px'
  }),
  indicatorsContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  valueContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  input: (styles: StylesConfig) => ({
    ...styles,
    marginLeft: '-2px',
    marginTop: '-10px',
    caretColor: '#999',
    boxShadow: 'none',
    border: 'none',
    maxHeight: '20px',
    minHeight: '20px',
    lineHeight: '12px'
  })
};

export const tagging = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  multiValue: (styles: StylesConfig) => {
    return {
      ...styles,
      maxHeight: 24,
      minHeight: 24,
      lineHeight: '17px',
      backgroundColor: 'hsl(0, 0%, 94%)',
      border: '1px solid hsl(0, 0%, 85%)'
    };
  }
};
