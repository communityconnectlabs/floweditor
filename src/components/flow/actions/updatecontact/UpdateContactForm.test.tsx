import UpdateContactForm from 'components/flow/actions/updatecontact/UpdateContactForm';
import { mock } from 'testUtils';
import {
  createSetContactFieldAction,
  getActionFormProps,
  createSetContactChannelAction,
  createSetContactNameAction,
  createSetContactLanguageAction
} from 'testUtils/assetCreators';
import * as utils from 'utils';
import * as React from 'react';
import { render, fireEvent } from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

// These are the RTL-based tests
describe(UpdateContactForm.name, () => {
  it('should render field', () => {
    const { baseElement } = render(
      <UpdateContactForm {...getActionFormProps(createSetContactFieldAction())} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render channel', () => {
    const { baseElement } = render(
      <UpdateContactForm {...getActionFormProps(createSetContactChannelAction())} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render name', () => {
    const { baseElement } = render(
      <UpdateContactForm {...getActionFormProps(createSetContactNameAction())} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render language', () => {
    const { baseElement } = render(
      <UpdateContactForm {...getActionFormProps(createSetContactLanguageAction())} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('shouldnt allow expressions in language selection', () => {
    const { getAllByDisplayValue } = render(
      <UpdateContactForm {...getActionFormProps(createSetContactLanguageAction())} />
    );

    // we don't support expressions but this shouldn't blow up either
    const input = getAllByDisplayValue('')[1];
    fireEvent.change(input, { target: { value: '@' } });
  });
});
