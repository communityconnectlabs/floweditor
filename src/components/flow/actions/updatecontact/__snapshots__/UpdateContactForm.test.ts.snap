// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateContactForm render should render an empty form with different action 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="set_contact_field"
  title="Update Contact"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "set_contact_name",
          "set_contact_language",
          "set_contact_channel",
        ],
        "component": [Function],
        "description": "Update the contact",
        "form": [Function],
        "name": "Update Contact",
        "type": "set_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    Select what to update
  </p>
  <AssetSelector
    additionalOptions={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
        Object {
          "id": "language",
          "name": "Language",
          "type": "property",
        },
        Object {
          "id": "channel",
          "name": "Channel",
          "type": "property",
        },
      ]
    }
    assets={
      Object {
        "items": Object {},
        "type": "field",
      }
    }
    createAssetFromInput={[Function]}
    createPrefix="Create Contact Field: "
    entry={
      Object {
        "value": Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      }
    }
    name="Contact Field"
    onAssetCreated={[Function]}
    onChange={[Function]}
    searchable={true}
    sortFunction={[Function]}
  />
  <div
    className="value"
  >
    <Connect(TextInputElement)
      autocomplete={true}
      entry={
        Object {
          "value": "",
        }
      }
      focus={true}
      name="Name"
      onChange={[Function]}
      onFieldFailures={[Function]}
      placeholder="Enter a new name for the contact"
    />
  </div>
</Dialog>
`;

exports[`UpdateContactForm render should render an empty form with no action 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="set_contact_field"
  title="Update Contact"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "set_contact_name",
          "set_contact_language",
          "set_contact_channel",
        ],
        "component": [Function],
        "description": "Update the contact",
        "form": [Function],
        "name": "Update Contact",
        "type": "set_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    Select what to update
  </p>
  <AssetSelector
    additionalOptions={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
        Object {
          "id": "language",
          "name": "Language",
          "type": "property",
        },
        Object {
          "id": "channel",
          "name": "Channel",
          "type": "property",
        },
      ]
    }
    assets={
      Object {
        "items": Object {},
        "type": "field",
      }
    }
    createAssetFromInput={[Function]}
    createPrefix="Create Contact Field: "
    entry={
      Object {
        "value": Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      }
    }
    name="Contact Field"
    onAssetCreated={[Function]}
    onChange={[Function]}
    searchable={true}
    sortFunction={[Function]}
  />
  <div
    className="value"
  >
    <Connect(TextInputElement)
      autocomplete={true}
      entry={
        Object {
          "value": "",
        }
      }
      focus={true}
      name="Name"
      onChange={[Function]}
      onFieldFailures={[Function]}
      placeholder="Enter a new name for the contact"
    />
  </div>
</Dialog>
`;

exports[`UpdateContactForm should allow switching from router to channel 1`] = `
Object {
  "channel": Object {
    "validationFailures": Array [],
    "value": Object {
      "id": "channel_id",
      "name": "Channel Name",
      "type": "channel",
    },
  },
  "field": Object {
    "value": Object {
      "id": "channel",
      "name": "Channel",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "value": "",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_channel",
  "valid": true,
}
`;

exports[`UpdateContactForm should allow switching from router to channel 2`] = `
Array [
  Object {
    "channel": Object {
      "name": "Channel Name",
      "uuid": "channel_id",
    },
    "type": "set_contact_channel",
    "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
  },
  [Function],
]
`;

exports[`UpdateContactForm should allow switching from router to contact name 1`] = `
Array [
  Object {
    "name": "Rowan Seymour",
    "type": "set_contact_name",
    "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
  },
  [Function],
]
`;

exports[`UpdateContactForm should allow switching from router to field value 1`] = `
Object {
  "channel": Object {
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "birthday",
      "name": "Birthday",
      "type": "field",
    },
  },
  "fieldValue": Object {
    "validationFailures": Array [],
    "value": "12/25/00",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_field",
  "valid": true,
}
`;

exports[`UpdateContactForm should allow switching from router to field value 2`] = `
Array [
  Object {
    "field": Object {
      "key": "birthday",
      "name": "Birthday",
    },
    "type": "set_contact_field",
    "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
    "value": "12/25/00",
  },
  [Function],
]
`;

exports[`UpdateContactForm should allow switching from router to language 1`] = `
Object {
  "channel": Object {
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "language",
      "name": "Language",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "value": "",
  },
  "language": Object {
    "validationFailures": Array [],
    "value": "e",
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_language",
  "valid": true,
}
`;

exports[`UpdateContactForm should allow switching from router to language 2`] = `
Array [
  Object {
    "language": undefined,
    "type": "set_contact_language",
    "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
  },
  [Function],
]
`;

exports[`UpdateContactForm updates should update channel 1`] = `
Object {
  "channel": Object {
    "validationFailures": Array [],
    "value": Object {
      "id": "channel_id",
      "name": "Channel Name",
      "type": "channel",
    },
  },
  "field": Object {
    "value": Object {
      "id": "channel",
      "name": "Channel",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "value": "25",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_channel",
  "valid": true,
}
`;

exports[`UpdateContactForm updates should update channel 2`] = `
Array [
  Object {
    "name": "Rowan Seymour",
    "type": "set_contact_name",
    "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
  },
  [Function],
]
`;

exports[`UpdateContactForm updates should update field value 1`] = `
Object {
  "channel": Object {
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "birthday",
      "name": "Birthday",
      "type": "field",
    },
  },
  "fieldValue": Object {
    "validationFailures": Array [],
    "value": "12/25/00",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_field",
  "valid": true,
}
`;

exports[`UpdateContactForm updates should update field value 2`] = `
Array [
  Object {
    "name": "Rowan Seymour",
    "type": "set_contact_name",
    "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
  },
  [Function],
]
`;

exports[`UpdateContactForm updates should update language 1`] = `
Object {
  "channel": Object {
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "language",
      "name": "Language",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "value": "25",
  },
  "language": Object {
    "validationFailures": Array [],
    "value": "e",
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_language",
  "valid": true,
}
`;

exports[`UpdateContactForm updates should update language 2`] = `
Array [
  Object {
    "name": "Rowan Seymour",
    "type": "set_contact_name",
    "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
  },
  [Function],
]
`;

exports[`UpdateContactForm updates should update name 1`] = `
Object {
  "channel": Object {
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "name",
      "name": "Name",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "validationFailures": Array [],
    "value": "",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "Rowan Seymour",
  },
  "type": "set_contact_name",
  "valid": true,
}
`;

exports[`UpdateContactForm updates should update name 2`] = `
Array [
  Object {
    "name": "Rowan Seymour",
    "type": "set_contact_name",
    "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
  },
  [Function],
]
`;

exports[`UpdateContactForm updates should validate before saving 1`] = `
Object {
  "channel": Object {
    "validationFailures": Array [
      Object {
        "message": "Channel is required",
      },
    ],
    "value": null,
  },
  "field": Object {
    "value": Object {
      "id": "channel",
      "name": "Channel",
      "type": "property",
    },
  },
  "fieldValue": Object {
    "value": "25",
  },
  "language": Object {
    "value": null,
  },
  "name": Object {
    "value": "",
  },
  "type": "set_contact_channel",
  "valid": false,
}
`;
